src/Compilers/VisualBasic/Portable/Symbols/Source/SourceModuleSymbol.vb

```diff
@@ -887,19 +887,27 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
                                                symbol As Symbol) As Boolean
             Debug.Assert(Me.DeclaringCompilation.EventQueue IsNot Nothing)
 
+            Dim change As Boolean = False
             SyncLock _diagnosticLock
-                Dim change = (variable And mask) = comparand
+                change = (variable And mask) = comparand
                 If change Then
-                    Me.DeclaringCompilation.SymbolDeclaredEvent(symbol)
-
                     If Not ThreadSafeFlagOperations.Set(variable, mask) Then
                         ' If this gets hit, then someone wrote to variable without going through this
                         ' routine, which is a bug.
                         Throw ExceptionUtilities.Unreachable
                     End If
                 End If
-                Return change
             End SyncLock
+
+            If change Then
+                ' Invoke the symbol declared event outside the lock.
+                ' Task scheduler might decide to inline the event processing, which can call back into analyzers,
```

Comment:
To be more precise, this is the code path being taken:
-> [SymbolDeclaredEvent](http://source.roslyn.io/Microsoft.CodeAnalysis.VisualBasic/R/d7a2310fa51fca81.html)
-> [EnqueueCore](http://source.roslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/AsyncQueue.cs,55)
-> [waiter.SetResult](http://source.roslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/AsyncQueue.cs,91)
-> [TrySetResult](http://referencesource.microsoft.com/#mscorlib/system/threading/Tasks/TaskCompletionSource.cs,321)
-> [Task.TrySetResult](http://referencesource.microsoft.com/#mscorlib/system/threading/Tasks/TaskCompletionSource.cs,298)
-> [FinishStageThree](http://referencesource.microsoft.com/#mscorlib/system/threading/Tasks/Future.cs,488)
-> [FinishContinuations](http://referencesource.microsoft.com/#mscorlib/system/threading/Tasks/Task.cs,2362)

Do you think this is a bug in the AsyncQueue?